File: RetroKS/Kernel/Kernel.vb

'    This file is part of RetroKS
'
'    RetroKS is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    RetroKS is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

'About this kernel: THIS KERNEL IS NEAR TO BETA BUT NOT FINAL! Final kernel will be developed through another language, ASM included, depending on system.

Imports System.IO
Imports System.Reflection

Module Kernel

    'Variables
    Public Hddsize As String                                                            'The size of simulated Hard Drive
    Public Dsize As String                                                              'Same as above, but in bytes
    Public Hddmodel As String                                                           'Model of hard drive
    Public Cpuname As String                                                            'CPU name
    Public Cpuspeed As String                                                           'CPU Clock Speed
    Public SysMem As String                                                             'Memory of simulated system
    Public BIOSCaption As String                                                        'BIOS Caption
    Public BIOSMan As String                                                            'BIOS Manufacturer
    Public BIOSSMBIOSVersion As String                                                  'BIOS Version from SMBIOS
    Public BIOSVersion As String                                                        'BIOS Version (some AMI BIOSes output "AMIINT - 10") 
    Public KernelVersion As String = Assembly.GetExecutingAssembly().GetName().Version.ToString()
    Public BootArgs() As String                                                         'Array for boot arguments
    Public AvailableArgs() As String = {"motd", "nohwprobe", "chkn=1", "preadduser", "hostname", "quiet", "gpuprobe", "cmdinject", "debug", "help"}
    Public availableCMDLineArgs() As String = {"createConf", "promptArgs"}              'Array for available command-line arguments
    Public slotsUsedName As String                                                      'Lists slots by names
    Public slotsUsedNum As Integer                                                      'Lists slots by numbers
    Public Capacities() As String                                                       'Capacity (in MB)
    Public totalSlots As Integer                                                        'Total slots
    Public configReader As StreamReader                                                 'Configuration file
    Declare Sub Sleep Lib "kernel32" (milliseconds As Integer)                    'Enable sleep (Mandatory, don't remove)

    Sub KernelMain()

        'A title
        Console.Title = "RetroKS version " & KernelVersion

        Try
            'Parse real command-line arguments
            For Each argu In Environment.GetCommandLineArgs
                parseCMDArguments(argu)
            Next

            'Make an app data folder
            If Not Directory.Exists(AppDataPath) Then Directory.CreateDirectory(AppDataPath)

            'Create config file and then read it
            checkForUpgrade()
            If File.Exists(AppDataPath + "\kernelConfig.ini") = True Then
                configReader = My.Computer.FileSystem.OpenTextFileReader(AppDataPath + "\kernelConfig.ini")
            Else
                createConfig(False)
                configReader = My.Computer.FileSystem.OpenTextFileReader(AppDataPath + "\kernelConfig.ini")
            End If
            readConfig()

            'Show introduction. Don't remove license.
            Wln("|--+---> Welcome to the kernel, version {0} <---+--|", "neutralText", KernelVersion)
            Wln(vbNewLine + "    RetroKS  Copyright (C) 2022  EoflaOE" + vbNewLine +
                            "    This program comes with ABSOLUTELY NO WARRANTY, not even " + vbNewLine +
                            "    MERCHANTABILITY or FITNESS for particular purposes." + vbNewLine +
                            "    This is free software, and you are welcome to redistribute it" + vbNewLine +
                            "    under certain conditions; See COPYING file in source code." + vbNewLine, "license")

            'Phase 0: Initialize time and files, and check for quietness
            If argsOnBoot = True Then
                PromptArgs()
                If argsFlag = True Then
                    ParseArguments()
                End If
            End If
            If argsInjected = True Then
                ParseArguments()
                answerargs = ""
                argsInjected = False
            End If
            If TimeDateIsSet = False Then
                InitializeTimeDate()
                TimeDateIsSet = True
            End If
            Init()
            Wdbg("Kernel initialized, version {0}.", True, KernelVersion)
            If Quiet = True Or quietProbe = True Then
                'Continue the kernel, and don't print messages
                'Phase 1: Username management
                initializeMainUsers()
                If enableDemo = True Then
                    adduser("demo")
                End If
                LoginFlag = True

                'Phase 2: Check for pre-user making
                If CruserFlag = True Then
                    adduser(arguser, argword)
                End If

                'Phase 3: Free unused RAM and log-in
                DisposeAll()
                If LoginFlag = True And maintenance = False Then
                    LoginPrompt()
                ElseIf LoginFlag = True And maintenance = True Then
                    LoginFlag = False
                    Wln("Enter the admin password for maintenance.", "neutralText")
                    answeruser = "root"
                    showPasswordPrompt("root")
                End If
            Else
                'Continue the kernel
                'Phase 1: Username management
                initializeMainUsers()
                If enableDemo = True Then
                    adduser("demo")
                End If
                LoginFlag = True

                'Phase 2: Check for pre-user making
                If CruserFlag = True Then
                    adduser(arguser, argword)
                End If

                'Phase 3: Free unused RAM and log-in if the kernel isn't on maintenance mode
                DisposeAll()
                If LoginFlag = True And maintenance = False Then
                    LoginPrompt()
                ElseIf LoginFlag = True And maintenance = True Then
                    LoginFlag = False
                    Wln("Enter the admin password for maintenance.", "neutralText")
                    answeruser = "root"
                    showPasswordPrompt("root")
                End If
            End If
        Catch ex As Exception
            If DebugMode = True Then
                Wln(ex.StackTrace, "uncontError") : Wdbg(ex.StackTrace, True)
            End If
            KernelError(CChar("U"), True, 5, "Kernel Error while booting: " + Err.Description)
        End Try

    End Sub

End Module

File: RetroKS/Kernel/Kernel.cs
//
// RetroKS  Copyright (C) 2022  Aptivi
//
// This file is part of RetroKS
//
// RetroKS is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// RetroKS is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY, without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

using System;
using System.IO;
using System.Reflection;

namespace RetroKS
{

    static class Kernel
    {

        // Variables
        public static string KernelVersion = Assembly.GetExecutingAssembly().GetName().Version.ToString();
        public static string[] BootArgs;
        public static string[] AvailableArgs =
            new string[] { "motd", "nohwprobe", "chkn=1", "preadduser", "hostname", "quiet", "gpuprobe", "cmdinject", "debug", "help" };
        public static string[] availableCMDLineArgs =
            new string[] { "createConf", "promptArgs" };
        public static string slotsUsedName;
        public static int slotsUsedNum;
        public static int totalSlots;
        public static StreamReader configReader;
        public static string Host;
        public static string MOTD;
        public readonly static string NewLine = Environment.NewLine;

        public static void KernelMain()
        {

            // A title
            Console.Title = "RetroKS version " + KernelVersion;

            try
            {
                // Parse real command-line arguments
                foreach (var argu in Environment.GetCommandLineArgs())
                    CommandLineArgsParse.parseCMDArguments(argu);

                // Make an app data folder
                if (!Directory.Exists(InitializeDirectoryFile.AppDataPath))
                    Directory.CreateDirectory(InitializeDirectoryFile.AppDataPath);

                // Create config file and then read it
                Config.checkForUpgrade();
                if (File.Exists(InitializeDirectoryFile.AppDataPath + @"\kernelConfig.ini") == true)
                {
                    configReader = new StreamReader(InitializeDirectoryFile.AppDataPath + @"\kernelConfig.ini");
                }
                else
                {
                    Config.createConfig(false);
                    configReader = new StreamReader(InitializeDirectoryFile.AppDataPath + @"\kernelConfig.ini");
                }
                Config.readConfig();

                // Show introduction. Don't remove license.
                TextWriterColor.Wln("|--+---> Welcome to the kernel, version {0} <---+--|", "neutralText", KernelVersion);
                TextWriterColor.Wln(NewLine + "    RetroKS  Copyright (C) 2022  Aptivi" + NewLine + "    This program comes with ABSOLUTELY NO WARRANTY, not even " + NewLine + "    MERCHANTABILITY or FITNESS for particular purposes." + NewLine + "    This is free software, and you are welcome to redistribute it" + NewLine + "    under certain conditions; See COPYING file in source code." + NewLine, "license");

                // Phase 0: Initialize time and files, and check for quietness
                if (Flags.argsOnBoot == true)
                {
                    ArgumentPrompt.PromptArgs();
                    if (Flags.argsFlag == true)
                    {
                        ArgumentParse.ParseArguments();
                    }
                }
                if (Flags.argsInjected == true)
                {
                    ArgumentParse.ParseArguments();
                    ArgumentPrompt.answerargs = "";
                    Flags.argsInjected = false;
                }
                if (Flags.TimeDateIsSet == false)
                {
                    TimeDate.InitializeTimeDate();
                    Flags.TimeDateIsSet = true;
                }
                InitializeDirectoryFile.Init();
                DebugWriter.Wdbg("Kernel initialized, version {0}.", true, KernelVersion);
                if (Flags.Quiet == true | Flags.quietProbe == true)
                {
                    // Continue the kernel, and don't print messages
                    // Phase 1: Username management
                    UserManagement.initializeMainUsers();
                    if (Flags.enableDemo == true)
                    {
                        UserManagement.adduser("demo");
                    }
                    Flags.LoginFlag = true;

                    // Phase 2: Check for pre-user making
                    if (Flags.CruserFlag == true)
                    {
                        UserManagement.adduser(ArgumentParse.arguser, ArgumentParse.argword);
                    }

                    // Phase 3: Free unused RAM and log-in
                    DisposeExit.DisposeAll();
                    if (Flags.LoginFlag == true & Flags.maintenance == false)
                    {
                        Login.LoginPrompt();
                    }
                    else if (Flags.LoginFlag == true & Flags.maintenance == true)
                    {
                        Flags.LoginFlag = false;
                        TextWriterColor.Wln("Enter the admin password for maintenance.", "neutralText");
                        Login.answeruser = "root";
                        Login.showPasswordPrompt("root");
                    }
                }
                else
                {
                    // Continue the kernel
                    // Phase 1: Username management
                    UserManagement.initializeMainUsers();
                    if (Flags.enableDemo == true)
                    {
                        UserManagement.adduser("demo");
                    }
                    Flags.LoginFlag = true;

                    // Phase 2: Check for pre-user making
                    if (Flags.CruserFlag == true)
                    {
                        UserManagement.adduser(ArgumentParse.arguser, ArgumentParse.argword);
                    }

                    // Phase 3: Free unused RAM and log-in if the kernel isn't on maintenance mode
                    DisposeExit.DisposeAll();
                    if (Flags.LoginFlag == true & Flags.maintenance == false)
                    {
                        Login.LoginPrompt();
                    }
                    else if (Flags.LoginFlag == true & Flags.maintenance == true)
                    {
                        Flags.LoginFlag = false;
                        TextWriterColor.Wln("Enter the admin password for maintenance.", "neutralText");
                        Login.answeruser = "root";
                        Login.showPasswordPrompt("root");
                    }
                }
            }
            catch (Exception ex)
            {
                if (Flags.DebugMode == true)
                {
                    TextWriterColor.Wln(ex.StackTrace, "uncontError");
                    DebugWriter.Wdbg(ex.StackTrace, true);
                }
                KernelTools.KernelError('U', true, 5L, "Kernel Error while booting: " + ex.Message);
            }

        }

    }
}
